src/compute/src/sink/correction_v2.rs

```diff
@@ -0,0 +1,1109 @@
+            return vec![self];
+        }
+
+        // Otherwise we need to split the cursor so that each new cursor only yields runs of
+        // updates that are correctly (time, data)-ordered when advanced by `since_ts`. We achieve
+        // this by splitting the cursor at each time <= `since_ts`.
+        let mut splits = Vec::new();
+        let mut remaining = Some(self);
+
+        while let Some(cursor) = remaining.take() {
+            let (_, time, _) = cursor.get();
+            if time >= since_ts {
+                splits.push(cursor);
+                break;
+            }
+
+            let mut current = cursor.clone();
+            if let Some((cursor, skipped)) = cursor.skip_time(time) {
+                remaining = Some(cursor);
+                current = current.set_limit(skipped).expect("skipped at least 1");
+            }
+            current.overwrite_ts = Some(since_ts);
+            splits.push(current);
+        }
+
+        splits
+    }
+
+    /// Split the cursor at the given time.
+    ///
+    /// Returns two cursors, the first yielding all updates at times < `time`, the second yielding
+    /// all updates at times >= `time`. Both can be `None` if they would be empty.
+    fn split_at_time(self, time: Timestamp) -> (Option<Self>, Option<Self>) {
+        let Some(skip_ts) = time.step_back() else {
+            return (None, Some(self));
+        };
+
+        let before = self.clone();
+        match self.skip_time(skip_ts) {
+            Some((beyond, skipped)) => (before.set_limit(skipped), Some(beyond)),
+            None => (Some(before), None),
+        }
+    }
+
+    /// Attempt to unwrap the cursor into a [`Chain`].
+    ///
+    /// This operation efficiently reuses chunks by directly inserting them into the output chain
+    /// where possible.
+    ///
+    /// An unwrap is only successful if the cursor's `limit` and `overwrite_ts` are both `None` and
+    /// the cursor has unique references to its chunks.
+    fn try_unwrap(self) -> Result<Chain<D>, (&'static str, Self)> {
+        if self.limit.is_some() {
+            return Err(("cursor with limit", self));
+        }
+        if self.overwrite_ts.is_some() {
+            return Err(("cursor with overwrite_ts", self));
+        }
+        if self.chunks.iter().any(|c| Rc::strong_count(c) != 1) {
+            return Err(("cursor on shared chunks", self));
+        }
+
+        let mut chain = Chain::default();
+        let mut remaining = Some(self);
+
+        // We might be partway through the first chunk, in which case we can't reuse it but need to
+        // allocate a new one to contain only the updates the cursor can still yield.
+        while let Some(cursor) = remaining.take() {
+            if cursor.chunk_offset == 0 {
+                remaining = Some(cursor);
+                break;
+            }
+            let update = cursor.get();
+            chain.push(update);
+            remaining = cursor.step();
+        }
```

Comment:
This loop here (and elsewhere) performs element-by-element copies. It seems to me that we could trade-off how compactly we store data with how much we need to copy around. For example, when we extracted some prefix of a chunk, we could just leave the remaining data in the chunk and remember the offset. Then, we'd need to add a check that chunks in a chain are always at least half full to avoid sitting on too much memory.

Reply:
Interesting! This PR takes the conservative approach of always packing updates as tightly as possible, in exchange for more work that has to be done. There are definitely ways to optimize things, but I'd like to leave these as follow-ups given that they make the code more complicated and this PR is already large enough.